BREAKING: integration of ADLS packages into one

begoldsm opened the pull request:
<!-- DO NOT DELETE THIS TEMPLATE -->

## Description
1. Integrate DataLake.StoreUploader into DataLake.Store
2. Add tests
3. Add wrapper methods to the datalake store filesystem.
4. Rename a lot of the classes to remove "upload only connotations"
(upload -> transfer)
5. Support portability in NetStandard1.5
6. Update package version to a preview package now that it is a breaking
change.
7. fixed and re-enabled flaky transfer tests.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/CONTRIBUTING.md).**
- [x] **The pull request does not introduce [breaking changes](https://github.com/Azure/azure-sdk-for-net/blob/dev/Documentation/breaking-changes.md).**

### [General Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as a link to the swagger spec, used to generate the code.
- [x] The `project.json` and `AssemblyInfo.cs` files have been updated with the new version of the SDK.

Commit:
BREAKING: integration of ADLS packages into one

1. Integrate DataLake.StoreUploader into DataLake.Store
2. Add tests
3. Add wrapper methods to the datalake store filesystem.
4. Rename a lot of the classes to remove "upload only connotations"
(upload -> transfer)
5. Support portability in NetStandard1.5
6. Update package version to a preview package now that it is a breaking
change.
7. fixed and re-enabled flaky transfer tests.

begoldsm commented:
@markcowl @cormacpayne  note that I am going to have a second commit that removes DataLake.StoreUploader from the repo so that the two commits are more manageable to look at. I will wait until this first one passes CI before sending through the other one.

Commit:
Refactor Upload -> Transfer where it makes sense

Since this is a huge breaking change anyway, it makes sense to have the
code and the names in the code properly reflect that this client can be
used for both upload and download, so removing references to "upload"
where we are doing both upload and download.

Commit:
Remove DataLake.StoreUploader project

This project has been integrated into DataLake.Store

matt1883 commented on src/ResourceManagement/DataLake.Store/Microsoft.Azure.Management.DataLake.Store/Customizations/DataTransfer/Common/SegmentTransferStatus.cs:
```diff
@@ -14,32 +14,40 @@
 
         /// <summary>
-        /// Indicates that the segment is currently being uploaded.
+        /// Indicates that the segment is currently being transfered.
```
Nit: should be 'transferred'

Commit:
Fix spelling of Transferred

markcowl commented:
@begoldsm  Please include the PR of the specs change, also please ensure that generate.cmd is using a reference to the commit containing you change (and not just a reference to the master baranch in the spec repo).

begoldsm commented:
@markcowl there are no auto generated code changes here. All changes are in customizations, so there is no swagger PR or swagger commit.

cormacpayne commented on src/ResourceManagement/DataLake.Store/Microsoft.Azure.Management.DataLake.Store/Customizations/DataTransfer/Common/TransferFolderMetadata.cs:
```diff
@@ -252,25 +252,45 @@ internal static UploadFolderMetadata LoadFrom(string filePath)
+                            int endIndex = startIndex + updatesPerThread;
+                            for (int j = startIndex; j < endIndex; j++)
+                            {
+                                if (j >= result.Files.Length)
```
@begoldsm 👍

cormacpayne commented on src/ResourceManagement/DataLake.Store/Microsoft.Azure.Management.DataLake.Store/Customizations/DataTransfer/DataLakeStoreTransferClient.cs:
```diff
@@ -529,6 +539,50 @@ private UploadMetadata GetMetadata()
 
+        private int GetDefaultPerFileThreadCount(long fileLength, long maxSegmentLength)
+        {
+            return (int)Math.Min(Math.Ceiling((double)fileLength / maxSegmentLength), DefaultMaxPerFileThreadCount);
```
@begoldsm could there ever a case where `maxSegmentLength` is zero?

begoldsm replied:
If the user specifies a max segment length of 0 we will fail earlier than here, but I can add that validation here as well.

cormacpayne commented on src/ResourceManagement/DataLake.Store/Microsoft.Azure.Management.DataLake.Store/Customizations/DataTransfer/DataLakeStoreTransferClient.cs:
```diff
@@ -529,6 +539,50 @@ private UploadMetadata GetMetadata()
+                concurrentFileCount = DefaultIdealConcurrentFileCount;
+            }
+
+            long avgFileSize = folderSize / fileCount;
```
@begoldsm same comment, but with `fileCount`

begoldsm replied:
agreed and fixed, we now throw if there are no files, since we can't upload nothing :)

cormacpayne commented on src/ResourceManagement/DataLake.Store/Microsoft.Azure.Management.DataLake.Store/Customizations/DataTransfer/DataLakeStoreTransferClient.cs:
```diff
@@ -1176,23 +1230,28 @@ private IProgress<SegmentUploadProgress> CreateSegmentProgressTracker(UploadMeta
                 (sup) =>
                 {
                     //update the overall progress and report it back
+                    if(_token.IsCancellationRequested)
```
@begoldsm should this `if-statement` be before the comment that precedes it?

cormacpayne reviewed (changes requested):
@begoldsm a few comments, let me know if you have any questions!

begoldsm commented on src/ResourceManagement/DataLake.Store/DataLakeStore.Tests/DataTransferTests/DataLakeStoreClientTests.cs:
```diff
@@ -120,83 +115,88 @@ public void DataLakeUploader_TargetExistsNoOverwrite()
         /// Tests the case of a fresh upload with multiple segments.
         /// </summary>
-        [Fact(Skip = "begoldsm has acknowledged the flakiness of these tests. Hence skipping and issue 2344 is opened to track the progress")]
+        [Fact]//(Skip = "begoldsm has acknowledged the flakiness of these tests. Hence skipping and issue 2344 is opened to track the progress")]
```
Good catch, yes I can!

begoldsm commented on src/ResourceManagement/DataLake.Store/DataLakeStore.Tests/DataTransferTests/MultipleSegmentUploaderTests.cs:
```diff
@@ -106,7 +106,7 @@ public void MultipleSegmentUploader_MultipleSegments()
         /// Tests an uneventful upload from scratch made of several segments
         /// </summary>
-        [Fact(Skip = "begoldsm has acknowledged the flakiness of these tests. Hence skipping and issue 2344 is opened to track the progress")]
+        [Fact]//(Skip = "begoldsm has acknowledged the flakiness of these tests. Hence skipping and issue 2344 is opened to track the progress")]
```
Same answer :)

begoldsm commented on src/ResourceManagement/DataLake.Store/DataLakeStore.Tests/DataTransferTests/MultipleSegmentUploaderTests.cs:
```diff
@@ -274,40 +274,40 @@ private void VerifyTargetStreamIsComplete(UploadSegmentMetadata segmentMetadata,
         {
             byte[] result = new byte[segment.Length];
-            Array.Copy(_smallFileContents, segment.SegmentNumber * metadata.SegmentLength, result, 0, segment.Length);
+            Array.Copy(_smallFileContents, (int)(segment.SegmentNumber * metadata.SegmentLength), result, 0, (int)segment.Length);
```
This is required because segmentLength is a long and you must explicitly down cast. There is no risk for loss of precision here, though, since the test ensures a very small file size.

begoldsm commented on src/ResourceManagement/DataLake.Store/DataLakeStore.Tests/DataTransferTests/DataLakeStoreClientTests.cs:
```diff
@@ -367,21 +391,21 @@ public void DataLakeUploader_ResumeUploadDownloadWithAllMissingFiles()
         /// Tests the resume upload when only some segments were uploaded previously with progress tracking enabled
         /// </summary>
-        [Fact(Skip = "begoldsm has acknowledged the flakiness of these tests.Hence skipping and issue 2344 is opened to track the progress")]
+        [Fact]//(Skip = "begoldsm has acknowledged the flakiness of these tests.Hence skipping and issue 2344 is opened to track the progress")]
```
same answer :)

begoldsm commented on src/ResourceManagement/DataLake.Store/Microsoft.Azure.Management.DataLake.Store/Customizations/FileSystemOperationsExtensions.Customizations.cs:
```diff
@@ -59,6 +59,5 @@ public static bool PathExists(this IFileSystemOperations operations, string acco
                 return _result.Body;
             }
         }
-
```
Stylecop was complaining about a blank line with no code after it. Is it important to add it back?

cormacpayne replied:
@begoldsm this is fine to keep 👍

begoldsm commented on src/ResourceManagement/DataLake.Store/Microsoft.Azure.Management.DataLake.Store/Customizations/DataTransfer/Common/TransferFolderMetadata.cs:
```diff
@@ -252,25 +252,45 @@ internal static UploadFolderMetadata LoadFrom(string filePath)
-                        Parallel.ForEach(result.Files, new ParallelOptions { MaxDegreeOfParallelism = -1 }, file =>
+                    // populate all child metadata file paths as well
+                    var localMetadataFolder = Path.GetDirectoryName(filePath);
+                    int updatesPerThread = (int)Math.Ceiling((double)result.Files.Length / threadCount);
```
It should never be, but I will add an explicit check.

begoldsm commented on src/ResourceManagement/DataLake.Store/DataLakeStore.Tests/DataTransferTests/DataLakeStoreClientTests.cs:
```diff
@@ -80,33 +81,27 @@ public void DataLakeUploader_InvalidParameters()
-                () => { new DataLakeStoreUploader(new UploadParameters(_largeFilePath, "1", "foo", perFileThreadCount: DataLakeStoreUploader.MaxAllowedThreads + 1, maxSegmentLength: 4 * 1024 * 1024), new InMemoryFrontEnd()); });
-
-            Assert.Throws<ArgumentOutOfRangeException>(
-                () => { new DataLakeStoreUploader(new UploadParameters(_largeFilePath, "1", "foo", concurrentFileCount: 0, maxSegmentLength: 4 * 1024 * 1024), new InMemoryFrontEnd()); });
```
Because they are tested elsewhere now. <= 0 thread counts now result in a default thread count being computed. Previously it would result in an error, so those tests are no longer valid and the new tests should validate this.

Commit:
address PR comments.

Add validation of maxSegmentLength
Add checks for 0 and negative numbers that should never be 0 or
negative.
Add tests for the same.

Commit:
Merge branch 'AutoRest' of https://github.com/Azure/azure-sdk-for-net into AutoRest
